### PR TITLE
qt5.qtwebengine: 5.15.2 -> 5.15.3

### PR DESCRIPTION
**Pull request by rnhmjoj**
###### Motivation for this change
Updating QtWebEngine for qutebrowser. This "release" includes lots of bug and security fixes.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/qt-5/5.15/default.nix [hunk 1]
```diff
@@ -48,6 +48,35 @@ let
...
+        src = fetchgit {
+          url = "https://github.com/qt/qtwebengine.git";
+          sha256 = "19l1i4sk3pvnwbvz5s97jchqawfz8k1xmjza29bgvkp1zz96r0jx";
+          inherit rev branchName;
+          leaveDotGit = true;
+        };
+
+        postUnpack = ''
+          # Set Chromium branch
+          git submodule set-branch --branch 87-based src/3rdparty
```
Comment (dotlambda) [inline pkgs/development/libraries/qt-5/5.15/default.nix:73]: Wouldn't an update to that branch make us have to change the hash?

**Review by dotlambda — CHANGES_REQUESTED** [review]
The commit message should say `pythonPackages`, not `pythonpackages`.

File: pkgs/development/libraries/qt-5/modules/qtwebengine.nix [hunk 2]
```diff
@@ -40,9 +49,17 @@ qtModule {
...
-    # Patch Chromium build tools
     ''
-      ( cd src/3rdparty/chromium; patchShebangs . )
+      # Patch Chromium build tools
+      (
+        cd src/3rdparty/chromium;
+
+        # Manually fix unsupported shebangs
+        substituteInPlace third_party/harfbuzz-ng/src/src/update-unicode-tables.make \
+          --replace "/usr/bin/env -S make -f" "/usr/bin/make -f" || true
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/qt-5/modules/qtwebengine.nix:59]: NixOS shouldn't contain /usr/bin/make either, right?

File: pkgs/development/libraries/qt-5/5.15/default.nix [hunk 3]
```diff
@@ -48,6 +48,35 @@ let
...
+          git submodule update
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/qt-5/5.15/default.nix:74]: Does this require network? This seems not very nixpkgs like I must admit.

File: pkgs/development/libraries/qt-5/5.15/default.nix [hunk 4]
```diff
@@ -48,6 +48,35 @@ let
...
+      in
+      {
+        version = "${branchName}-${lib.substring 0 8 rev}";
+
+        src = fetchgit {
+          url = "https://github.com/qt/qtwebengine.git";
+          sha256 = "19l1i4sk3pvnwbvz5s97jchqawfz8k1xmjza29bgvkp1zz96r0jx";
+          inherit rev branchName;
+          leaveDotGit = true;
+        };
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/qt-5/5.15/default.nix:69]: Please use fetchFromGitHub.

File: pkgs/development/libraries/qt-5/modules/qtwebengine.nix (hunk 2 above)
Comment (rnhmjoj) [reply pkgs/development/libraries/qt-5/modules/qtwebengine.nix:59]: It's just so `patchShebangs` will fix it later.

File: pkgs/development/libraries/qt-5/5.15/default.nix (hunk 4 above)
Comment (rnhmjoj) [reply pkgs/development/libraries/qt-5/5.15/default.nix:69]: It doesn't support `branchName`, besides it's just calling fetchgit indirectly in this case.

**Comment by rnhmjoj** [comment]
I should have checked more carefully: the whole `postUnpack` is being ignored by `qtModule`. I'll have to move it under qtwebengine.nix behind a `(version > "5.15.2")` gate.

File: pkgs/development/libraries/qt-5/5.15/default.nix (hunk 3 above)
Comment (rnhmjoj) [reply pkgs/development/libraries/qt-5/5.15/default.nix]: I opted to leave the submodule as is: it's already pointing to the right chromium branch, at least for now.

**Comment by petabyteboy** [comment]
I built the version before the force-push and it looked good:

```
Backend: QtWebEngine 5.15.3, Chromium 87.0.4280.144
```

Now building from the new commit.

**Comment by petabyteboy** [comment]
912ea22 does not build:

```
/build/qtwebengine-a059e74/src/core/api/qtwebenginecoreglobal_p.h:54:10: fatal error: QtWebEngineCore/qtwebenginecoreglobal.h: No such file or directory
   54 | #include <QtWebEngineCore/qtwebenginecoreglobal.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
[24358/25112] CXX obj/third_party/blink/renderer/platform/loader/loader/loader_jumbo_4.o
In file included from gen/third_party/blink/renderer/platform/loader/loader_jumbo_4.cc:10:
./../../3rdparty/chromium/third_party/blink/renderer/platform/loader/fetch/resource_load_scheduler.cc: In member function 'void blink::ResourceLoadScheduler::ShowConsoleMessageIfNeeded()':
./../../3rdparty/chromium/third_party/blink/renderer/platform/loader/fetch/resource_load_scheduler.cc:437:18: warning: variable 'target_option' set but not used [-Wunused-but-set-variable]
  437 |   ThrottleOption target_option;
      |                  ^~~~~~~~~~~~~
[24360/25112] CXX obj/third_party/blink/renderer/platform/platform/platform_jumbo_60.o
ninja: build stopped: subcommand failed.
make[3]: *** [Makefile.gn_run:370: run_ninja] Error 1
make[3]: Leaving directory '/build/qtwebengine-a059e74/src/core'
make[2]: *** [Makefile:82: sub-gn_run-pro-make_first] Error 2
make[2]: Leaving directory '/build/qtwebengine-a059e74/src/core'
make[1]: *** [Makefile:80: sub-core-make_first] Error 2
make[1]: Leaving directory '/build/qtwebengine-a059e74/src'
make: *** [Makefile:49: sub-src-make_first] Error 2
builder for '/nix/store/9hs1v19x014zmqml63fr88qsm4q11kl5-qtwebengine-5.15.3-a059e740.drv' failed with exit code 2
```

**Comment by rnhmjoj** [comment]
It was almost done: 24360/25112, damn.
The only differences are the patch and .git which is now removed.
I'll try reverting both changes but it will take a while.

File: pkgs/development/libraries/qt-5/5.15/default.nix (hunk 1 above)
Comment (petabyteboy) [reply pkgs/development/libraries/qt-5/5.15/default.nix]: I was wondering about this as well. Are we sure it will always give the same result?

**Comment by rnhmjoj** [comment]
It looks like it's `leaveDotGit`.
I have no idea why this could have anything to do with the missing header.

**Comment by peterhoeg** [comment]
> I have no idea why this could have anything to do with the missing header.

Is there a script that runs early that calls git in order to create a header file which now fails without a proper repository?

**Comment by wedens** [comment]
@rnhmjoj Since you're bumping qutebrowser version, could you add a `pynacl` dependency for the new `qute-keepassxc` userscript?

That's unfortunate that there is no way to add a userscript dependency without modifying qutebrowser's derivation, but that's another story...

File: pkgs/development/libraries/qt-5/5.15/default.nix (hunk 4 above)
Comment (SuperSandro2000) [reply pkgs/development/libraries/qt-5/5.15/default.nix]: Why do we require the branch name? revs are unique over all branches.

**Comment by rnhmjoj** [comment]
> Is there a script that runs early that calls git in order to create a header file which now fails without a proper repository?

I spent some time trying to test this theory. It seems git is indeed needed (ie. build fails if git is replaced by a stub command), but I couldn't figure how or where is used.

> @rnhmjoj Since you're bumping qutebrowser version, could you add a pynacl dependency for the new qute-keepassxc userscript?

Done. I tried running the script and it seems to work.

**Comment by rnhmjoj** [comment]
branchName sets the name of checked out branch, otherwise fetchgit will
rename it to "fetchgit". I think the build system may access it to get
the version/branch info, so it's better to keep the real name.

**Comment by rnhmjoj** [comment]
I've built and tested qutebrowser, webmacs and falkon. Each seems to be working fine, also the patch has fixed the locales crash. Can anybody run a nixpkgs-review to check the rest of the packages?

Is this ok for master or should this go through staging?

**Comment by FRidh** [comment]
master is fine, there is a mass rebuild anyway

**Comment by vcunat** [comment]
There's a [problem](https://hydra.nixos.org/build/140311701):
> Output limit exceeded

i.e. the produced size is over the 2 GiB limit we have on Hydra.

**Comment by vcunat** [comment]
I forgot to mention &ndash; this problem will be blocking the `nixos-unstable` channel, though fortunately not the first version with updated openssl.

**Comment by rnhmjoj** [comment]
Uhm, the source tarball is huge. It's probably due to leaveDotGit, even if it strips down the repository.

**Comment by vcunat** [comment]
Ah, right, I didn't notice it's about the `src` derivation already.

**Comment by rnhmjoj** [comment]
It looks like the problem is not the top-level .git but the one from the `3rdparty` submodule, which hasn't been stripped down: it's around 1.5GB. Deleting this would bring the size down to 2.2GB, which is still too much? Is the limit 2GB?

**Comment by vcunat** [comment]
Yes, I believe it's 2^31 NAR bytes.  You can locally get the number e.g. by `nix path-info -S /path`.

**Comment by dotlambda** [comment]
This broke QtWebChannel:
```
CMake Error at /nix/store/bfv70pjhrmpbxmavm7x20wzrdari48mw-qtwebengine-5.15.3-a059e74-dev/lib/cmake/Qt5WebEngineCore/Qt5WebEngineCoreConfig.cmake:97 (find_package):
  Could not find a configuration file for package "Qt5WebChannel" that is
  compatible with requested version "5.15.3".

  The following configuration files were considered but not accepted:

    /nix/store/i33h0wqanlbvjpcfianvvf72v1l94407-qtwebchannel-5.15.2-dev/lib/cmake/Qt5WebChannel/Qt5WebChannelConfig.cmake, version: 5.15.2
```

**Comment by petabyteboy** [comment]
I guess we will need something like this: https://github.com/archlinux/svntogit-packages/commit/1f64d05d0217a396888062dd420bd77af8b49e77#diff-37538beb61ff63edebbf735dfcf39e5d732f49183d6beb097169d971875ca422R63

**Comment by doronbehar** [comment]
Today I noticed that `nextcloud-client` fails to build because it needs qtquickcontrols 5.15.3 and no earlier due to qtwebengine, very similar to what @dotlambda reported:

```
nextcloud-client> CMake Error at /nix/store/bfv70pjhrmpbxmavm7x20wzrdari48mw-qtwebengine-5.15.3-a059e74-dev/lib/cmake/Qt5WebEngineCore/Qt5WebEngineCoreConfig.cmake:97 (find_package):
nextcloud-client>   Could not find a configuration file for package "Qt5Quick" that is
nextcloud-client>   compatible with requested version "5.15.3".
nextcloud-client>   The following configuration files were considered but not accepted:
nextcloud-client>     /nix/store/lg0gwblna9njvb8alv8rhljqm52q85z0-qtdeclarative-5.15.2-dev/lib/cmake/Qt5Quick/Qt5QuickConfig.cmake, version: 5.15.2
nextcloud-client> Call Stack (most recent call first):
nextcloud-client>   /nix/store/bfv70pjhrmpbxmavm7x20wzrdari48mw-qtwebengine-5.15.3-a059e74-dev/lib/cmake/Qt5WebEngineWidgets/Qt5WebEngineWidgetsConfig.cmake:97 (find_package)
nextcloud-client>   /nix/store/5s812h9jxlqmwyx3x4abs74167hqnm8p-qtbase-5.15.2-dev/lib/cmake/Qt5/Qt5Config.cmake:28 (find_package)
nextcloud-client>   src/CMakeLists.txt:7 (find_package)
nextcloud-client> -- Configuring incomplete, errors occurred!
nextcloud-client> See also "/build/source/build/CMakeFiles/CMakeOutput.log".
```

Why did you update only qtwebengine and not all of qt5??

**Comment by dotlambda** [comment]
@doronbehar Should have been fixed by https://github.com/NixOS/nixpkgs/pull/118084

QT 5.15.3 is only available under a commercial license: https://www.qt.io/blog/commercial-lts-qt-5.15.3-released

**Pull request merged**